### ISSUE TITLE
'--once' option dosen't work well with 'docker exec -it /bin/bash' 

### ISSUE THREAD

**Issue by tczf1128** [yudai/gotty#70]
It dosen't exit , when I set gotty to a daemon process use nohup or daemon command .

**Comment by yudai** [comment]
Hi, could you provide the actual commands you tried? `gotty --once -w docker run -it --rm ubuntu /bin/bash` worked properly on my environment.

**Comment by tczf1128** [comment]
@yudai   `daemon  --name=gotty  -U  -o /home/gotty.log -P /home/ -- /usr/sbin/gotty -r -w --once docker exec -it 942d3cd41cc6 /bin/bash`
thanks!

**Comment by yudai** [comment]
It looks working well on my env. Actually, gotty stops working after a client exits the bash process. No additional client cannot connect to gotty after that.
I assume what you want to archive requires the `-r` option of the `daemon` command. It restarts a new process once a child process exits. With this option, `daemon` restarts a new gotty process after the previous gotty client disconnected and the new gotty process has a new random URL.

**Comment by tczf1128** [comment]
It stops working and no additional client cannot connect to gotty, but the process of gotty still exist

**Comment by yudai** [comment]
Did you try with the `-r` option of the `daemon` command?

I confirmed that the process of gotty disappears after the first client exits when I don't provide the `-r` option.

**Comment by tczf1128** [comment]
No.  My command is `daemon  --name=gotty  -U -o /home/log -F /home/pid -- /usr/sbin/gotty -r -w --once docker exec -it 3d7f353e784e /bin/bash`
I want the gotty process exit when I close the client. It works for you?

**Comment by yudai** [comment]
Ok, I confirmed when I close the browser window (not running `exit`), gotty gets stuck with the log `websocket: close 1001`.
I'll fix the issue soon.

**Comment by tczf1128** [comment]
thank you

**Comment by yudai** [comment]
Confirmed that docker command ignores SIGHUP when it's launched by gotty. However, I have no idea why the process ignores the signal. Indeed, it accepts SIGTERM and quits. I need some more time to investigate the root cause of the problem.

**Comment by yudai** [comment]
Got a little bit progress. When ran `daemon --name=gotty -U -o /home/gotty.log -P /home/ -- /usr/sbin/gotty -r -w --once docker exec -it 942d3cd41cc6 top`, the process exited normally. So it seems like bash ignores SIGHUP in some circumstance?

**Comment by moul** [comment]
Did you tried to call a `SIGKILL` after the current `SIGHUP` https://github.com/yudai/gotty/blob/6013e439336db11fb0bcede9bfbd7e912cdfa435/app/client_context.go#L78 ?

**Comment by yudai** [comment]
@moul I added a new option that configures the signal when gotty kills child processes.

Indeed, `SIGKILL` always works, however I'm still wondering why only `docker exec bash` won't exit. There must be a condition that docker or bash ignores `SIGHUP`.

**Comment by moul** [comment]
Nohup/daemon are usually responsible of this kind of behavior by explicitly ignoring SIGHUP signals.
But in this context, docker is *behind* GoTTY, so the signals between the 2 processes should continue to pass

```
+------+   +--------------+ +-------+  +-------------+
|      |   |              | |       |  |             |
| bash |   | nohup/daemon | | GoTTY |  | docker exec |
|      |   |              | |       |  |             |
+---+--+   +--------------+ ++----+-+  +----+--------+
    |                        ^    |         ^         
    |                        |    |         |         
    +------------------------+    +---------+         
            SIGHUP ignored       SIGHUP should pass   
```

**Comment by yudai** [comment]
@moul Yes, that's the interesting point. I try sending SIGHUP directory to the `docker exec` process from another terminal, but it didn't work either. Mysterious...

**Comment by upccup** [comment]
@yudai  It issues new progress?  I have this problem too, but i found that when i closed this websocket  gotty code was block at https://github.com/yudai/gotty/blob/master/app/client_context.go#L74 and the processSend not return is this the question

**Comment by yudai** [comment]
I think this issue is resolved in #159. @upccup 's hint was helpful. Thanks!

**Comment by yudai** [comment]
Should be fixed in the master branch. Let me know when it still exists.
